base/applications/mspaint/history.cpp

```diff
@@ -209,6 +211,46 @@ void ImageModel::InvertColors()
     NotifyImageChanged();
 }
 
+/* This function only converts image to Grayscale without modify color depth */ 
+void ImageModel::GSColors()
+{
+    int x, y;
+    COLORREF pixel_color;
+
+    CopyPrevious();
+
+    for (x = 0; x < GetWidth(); x++)
+    {
+        for (y = 0; y < GetHeight(); y++)
+        {
+            pixel_color = GetPixel(hDrawingDC, x, y);
+            pixel_color = (pixel_color & 0xFF) * 0x10101;
+            SetPixel(hDrawingDC, x, y, pixel_color);
+        }
+    }
+    NotifyImageChanged();
+}
+
+void ImageModel::BKColors()
+{
+    int x, y;
+    COLORREF pixel_color;
+
+    CopyPrevious();
+
+    for (x = 0; x < GetWidth(); x++)
+    {
+        for (y = 0; y < GetHeight(); y++)
+        {
+            pixel_color = GetPixel(hDrawingDC, x, y);
+            if ((pixel_color & 0x00FFFFFF) != 0x00FFFFFF)
+                pixel_color = 0x00;
+            SetPixel(hDrawingDC, x, y, pixel_color);
+        }
+    }
```

Comment:
If the destination is going to be a genuine 1BPP bitmap, wouldn't the canonical solution use a simple BitBlt from one bitmap to the other?